### PR TITLE
Adds partitioning library

### PR DESCRIPTION
**Pull request by davidporter-id-au**
Adds partitioning library:

- Adds an interface for handling workflow isolation group redirection and failover
- Adds an interface for handling drain states for both global and domain-based drains
- Adds relevant internal types
- Adds a simple interface for handling global drains as stored in database-based dynamic config
- Adds a default implementation for the open-source implementation and implementation example

### CODE REVIEW

File: common/persistence/isolationGroupDrains.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"context"
+
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/types"
+)
+
+//go:generate mockgen -package $GOPACKAGE -source $GOFILE -destination zoneDrains_mock.go -self_package github.com/uber/cadence/common/persistence
+
+type GlobalIsolationGroupDrains interface {
```
Comment (Shaddoll) [inline common/persistence/isolationGroupDrains.go:34]: I don't think this interface is needed. We're using the dynamic config store to persist the drain config and should use the dynamic client to write and read the config.

File: common/types/partition.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package types
+
+const (
+	IsolationGroupStatusInvalid IsolationGroupStatus = iota
```
Comment (Shaddoll) [inline common/types/partition.go:26]: I think these definitions should be moved to `partition` package.
The `types` package should only contains the types defined in `idl`.

File: common/partition/default-partitioner.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+// where the workflow was started, and is expected to be pinned, and a RunID for a fallback means
+// to partition data deterministically.
+type DefaultWorkflowPartitionConfig struct {
+	WorkflowStartIsolationGroup types.IsolationGroupName
+	RunID                       string
+}
+
+// DefaultPartitioner is a business-agnostic implementation of partitioning
+// which is used by the Cadence system to allocate workflows in matching by isolation-group
+type DefaultPartitioner struct {
```
Comment (Shaddoll) [inline common/partition/default-partitioner.go:53]: I think this should be unexported

File: common/persistence/isolationGroupDrains.go (hunk 1 above)
Comment (davidporter-id-au) [reply common/persistence/isolationGroupDrains.go:34]: yes, this is for dynamicConfig, it's intended to use that in the DB with a different config type https://sourcegraph.uberinternal.com/github.com/uber/cadence/-/blob/common/persistence/dataManagerInterfaces.go#L224

File: common/types/partition.go (hunk 2 above)
Comment (davidporter-id-au) [reply common/types/partition.go:26]: ok! no worries

File: common/partition/partitioning.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	allIsolationGroups                []types.IsolationGroupName
+}
+
+// State is a convenience return type of a collection of IsolationGroup configurations
+type State struct {
+	Global types.IsolationGroupConfiguration
+	Domain types.IsolationGroupConfiguration
+}
+
+type Partitioner interface {
```
[the remark this answers is not in the document]
Comment (davidporter-id-au) [reply common/partition/partitioning.go:48]: let's sync to understand this

File: common/partition/default-isolation-group-state.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	"context"
+	"fmt"
+
+	"github.com/uber/cadence/common/persistence"
+
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+)
+
+type DefaultIsolationGroupStateHandler struct {
```
[the remark this answers is not in the document]
Comment (davidporter-id-au) [reply common/partition/default-isolation-group-state.go:35]: sure!

File: common/persistence/isolationGroupDrains.go (hunk 1 above)
Comment (davidporter-id-au) [reply common/persistence/isolationGroupDrains.go]: removed as per discussion, this will live at a higher-level in `common/isolation-groups`

**Comment by davidporter-id-au** [comment]
Closing this because I've apparently messed up the diff in favour of https://github.com/uber/cadence/pull/5169

**Pull request closed without merging**